### PR TITLE
236/solvers details

### LINKED ISSUE #236: Solvers Details page
**Description**
Create the Solvers Details page as specified on the designs

image

### PR DESCRIPTION
**Pull request by matextrem**
# Summary

Closes #236 

Added Solver details page.

Screen Shot 2022-10-21 at 01 20 55

# To Test

1.  Open the page `solvers`
2. Click on any solver name either in Active solvers or Settlements table
   * You'll be redirected to the solver details page.

*Disclaimer: There are some info left we have to update such as ETH cost and Active since box (it is harcoded right now)*

### CODE REVIEW

**Comment by elena-zh** [comment]
Hey @matextrem , nice changes!
List of issues from my side: 
1. Full date on the Active since card does not match the format we use for full dates across the app
вфеу ащкьфе
2. There is no Solver name when open the page from Settlements tab
name
3. Frankly, it is not clear to me why we show full amounts on cards when press on Tooltips icons.
image
 We show amounts on hover, and tooltips icons should open explanation modals for this or that field 
image

4. I think, it would be great to show a better message when a request to load the page data fails (error loading data, please try again?)
failed - error
5. When Trades amount is not loaded, it shows NaN instead of 0
nan when no data
6. When open Solver details page from Active solvers tab in Ethereum, all dates are in June 
june
7. Ethereum + Active solvers tab: the last solver in the list does not have a name, so the name is not clickable
no name - address

9. There are no names for solvers in Gorli
gorli no names
10. It is impossible to load data for single solvers page from Active Solvers tab in Gnosis chain
image
11. Some token icons (like GNO) are dark
dark tokens
12. Text on cards should be better aligned in a tablet/mobile view
cards better align
13. Name and a solver address are not aligned in a real iOS device
mobile

**Review by elena-zh — CHANGES_REQUESTED** [review]
See the notes above

**Comment by matextrem** [comment]
hey @elena-zh  I addressed some of the bugs you shared.

Regarding 1 & 3 it needs to be changed. I left it is since we need to discuss what to put in there. maybe @alongoni could help us in it.

4. We have a notification in place when there is an error. Those errors you are seeing are related to the icons not able to being loaded but it is not related to the solver details info itself.

6. Is that wrong? I opened up another solver and it showed me dates in May. cc: @ramirotw 

7. There is no name for that solver. Now, we are showing the address instead.

8. I didn't find the SQL query to fetch solvers name for Gorli. I tried to use the same we are using for Ethereum but the addresses don't match. cc: @alfetopito 

9. Yes, we are not getting any info related settlements for Gnosis chain (Subgraph is only returning the ids). cc @ramirotw 

10. It was fixed in this [PR](https://github.com/cowprotocol/explorer/pull/235). When merged, it will be reflected in this PR as well.

File: src/hooks/useGetSettlements.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  const fetchSettlements = useCallback(
+    async (network: Network): Promise<void> => {
+      setIsLoading(true)
+      setSettlements([])
+      try {
+        const response = await COW_SDK.cowSubgraphApi.runQuery<{
+          settlements: Settlement[]
+        }>(
+          solverAddress ? GET_SETTLEMENTS_BY_SOLVER_QUERY : GET_SETTLEMENTS_QUERY,
+          solverAddress ? { solver: solverAddress } : undefined,
```
Comment (ramirotw) [inline src/hooks/useGetSettlements.ts:28]: `solverAddress` should be lowercased, the subgraph doesn't work with checksummed addresses 

image

**Comment by ramirotw** [comment]
> When open Solver details page from Active solvers tab in Ethereum, all dates are in June

@matextrem @elena-zh This is because that solver was deactivated but the subgraph currently is not handling the toggle of the isSolver flag to false. There's a pending PR in https://github.com/cowprotocol/subgraph/pull/65 to fix that. 

To validate whether a mainnet solver is active or not using onchain data:

1. Go to https://etherscan.io/address/0x2c4c28ddbdac9c5e7055b4c863b72ea0149d8afe#readProxyContract
2. In the `isSolver` section, paste the solver address
3. If the `bool` response is `true` then it's active, if empty it's not

**Comment by ramirotw** [comment]
> It is impossible to load data for single solvers page from Active Solvers tab in Gnosis chain

@matextrem this issue is related to https://github.com/cowprotocol/explorer/pull/241#discussion_r1002167320

**Comment by elena-zh** [comment]
> > When open Solver details page from Active solvers tab in Ethereum, all dates are in June
> 
> @matextrem @elena-zh This is because that solver was deactivated but the subgraph currently is not handling the toggle of the isSolver flag to false. There's a pending PR in [cowprotocol/subgraph#65](https://github.com/cowprotocol/subgraph/pull/65) to fix that.
> 
> To validate whether a mainnet solver is active or not using onchain data:
> 
> 1. Go to https://etherscan.io/address/0x2c4c28ddbdac9c5e7055b4c863b72ea0149d8afe#readProxyContract
> 2. In the `isSolver` section, paste the solver address
> 3. If the `bool` response is `true` then it's active, if empty it's not

That is ridiculous :)
The tab name is 'Active solvers'. but we show deactivated solvers at the top of the table.

**Comment by elena-zh** [comment]
Hey @matextrem , thanks for the explanation. 
Here are my notes: 
1. For the cases 1, 3, agree, let waif for @alongoni input here
2. 8 case `There are no names for solvers in Gorli`: why can't we show solvers addresses instead? 
3. 9 case: `'It is impossible to load data for single solvers page from Active Solvers tab in Gnosis chain'`: I believe, it is going to be fixed
4. 11 case `Text on cards should be better aligned in a tablet/mobile view`: still relevant
mobile cards
5. 12 case: looks much better, but still the address is not placed in 1 line with a solver name
not in 1 line

Thank you!

**Comment by ramirotw** [comment]
> That is ridiculous :)
> The tab name is 'Active solvers'. but we show deactivated solvers at the top of the table.

Indeed it's ridiculous, but this is for internal testing before @GabrielCamba deploys the changes to fix it.

**Comment by alfetopito** [comment]
Looking good so far.

Comments about current progress:

1. Invert the position of `timestamp` and `tx hash` columns
Screen Shot 2022-10-25 at 16 45 12
2. For the tooltips in the solvers detail page, I'm inclined to either: Not show the tool tip OR put a brief and somewhat redundant explanation like
 - Trades: Total trades settled by this solver
 - Total volume: Total volume settled by this solver
 - Total settlements: Total settlements by this solver
 - Active since: When the solver became active as was able to submit solutions

3. For the `Trades` card, add `Total` => `Total trades`


And I have a question: are we also limiting the # of settlements we load per individual solver to 1000 like it's done in the general settlements page?

**Comment by alfetopito** [comment]
Request for a follow up PR:

Update the overall search so when a solver address is searched it leads to the solver details page.

For example. Searching on https://pr241--explorer.review.gnosisdev.com/gc for `0x7942a2b3540d1eC40B2740896f87aEcB2a588731` (Quasimodo's address) should lead you to https://pr241--explorer.review.gnosisdev.com/gc/solvers/0x7942a2b3540d1eC40B2740896f87aEcB2a588731 rather than https://pr241--explorer.review.gnosisdev.com/gc/address/0x7942a2b3540d1eC40B2740896f87aEcB2a588731

File: src/apps/explorer/components/SettlementsTableWidget/index.tsx [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  query: string
+  tableState: TableState
+  data: Settlement[]
+  setTableValues: (data: unknown) => void
+}
+
+export const SettlementsTableWidget: React.FC<Props> = ({ query, tableState, setTableValues, data }) => {
+  const networkId = useNetworkId() || undefined
+  const { settlements, isLoading, error } = useGetSettlements(networkId, data)
+  const filteredSettlements = useFlexSearch(query, settlements, ['name', 'txHash'])
```
Comment (alfetopito) [inline src/apps/explorer/components/SettlementsTableWidget/index.tsx:46]: Can the search have some debounce?

**Comment by elena-zh** [comment]
> Update the overall search so when a solver address is searched it leads to the solver details page.
> 
> For example. Searching on https://pr241--explorer.review.gnosisdev.com/gc for `0x7942a2b3540d1eC40B2740896f87aEcB2a588731` (Quasimodo's address) should lead you to https://pr241--explorer.review.gnosisdev.com/gc/solvers/0x7942a2b3540d1eC40B2740896f87aEcB2a588731 rather than https://pr241--explorer.review.gnosisdev.com/gc/address/0x7942a2b3540d1eC40B2740896f87aEcB2a588731

Hey @alfetopito , I have created #244 task for this

**Comment by elena-zh** [comment]
Hey @matextrem , cases 1, 2, 4, 5 from this [comment ](https://github.com/cowprotocol/explorer/pull/241#issuecomment-1288846978)are still not fixed.

Besides, I have another question: since the number of trades is limited with 1000 now, should we show this card on the UI? In most of cases it will always show 1K until we implement a solution how to show more data
will be always 1k

**Comment by ramirotw** [comment]
> Hey @matextrem , cases 1, 2, 4, 5 from this [comment ](https://github.com/cowprotocol/explorer/pull/241#issuecomment-1288846978)are still not fixed.
> 
> Besides, I have another question: since the number of trades is limited with 1000 now, should we show this card on the UI? In most of cases it will always show 1K until we implement a solution how to show more data
> will be always 1k
> 
> 

FYI the Total Settlements number will be updated once the subgraph update is deployed, hopefully by the EOW.

**Comment by matextrem** [comment]
> Looking good so far.
> 
> Comments about current progress:
> 
> 1. Invert the position of `timestamp` and `tx hash` columns
>    Screen Shot 2022-10-25 at 16 45 12
> 2. For the tooltips in the solvers detail page, I'm inclined to either: Not show the tool tip OR put a brief and somewhat redundant explanation like
> 
> * Trades: Total trades settled by this solver
> * Total volume: Total volume settled by this solver
> * Total settlements: Total settlements by this solver
> * Active since: When the solver became active as was able to submit solutions
> 
> 3. For the `Trades` card, add `Total` => `Total trades`
> 
> And I have a question: are we also limiting the # of settlements we load per individual solver to 1000 like it's done in the general settlements page?

Yes, we are limiting the amount of settlements to 1000 as well.

**Comment by matextrem** [comment]
> Hey @matextrem , cases 1, 2, 4, 5 from this [comment ](https://github.com/cowprotocol/explorer/pull/241#issuecomment-1288846978)are still not fixed.
> 
> Besides, I have another question: since the number of trades is limited with 1000 now, should we show this card on the UI? In most of cases it will always show 1K until we implement a solution how to show more data will be always 1k

Hey @elena-zh!  I updated the branch with the feedback.

Regarding point 2, It's working for me. 

Regarding 5, I don't fully understand what's the behavior you want here. They are vertically centered on Tablet/Mobile view : 

Screen Shot 2022-10-26 at 14 56 58

Maybe @alongoni could give feedback on this as well.

**Comment by elena-zh** [comment]
Hey @matextrem , overall, changes look good to me. But: 

> Regarding point 2, It's working for me.

Please, take a look at the Settlements tab. Active solvers tab looks good. 
image

Thanks!

**Comment by matextrem** [comment]
@elena-zh updated!

**Review by elena-zh — APPROVED** [review]
Hey @matextrem , changes LGTM now. 
We still have issues related to the Supgraph: 
 - Total settlements are 1000 for a single solver, and on Settlements table
 - No ETH cost
 - Do 'Active since'

Also, I have created a separate task to adjust pagination displaying on tables across the whole app #246. 

Thanks

**Pull request merged**